### PR TITLE
KOGITO-1373 Integrate the new StateControl API (Undo/Redo/Is Dirty) on the different channels

### PR DESCRIPTION
**Pull request by ljmotta**
## Tasks
 - https://issues.redhat.com/browse/KOGITO-1390
 - https://issues.redhat.com/browse/KOGITO-2169
 - https://issues.redhat.com/browse/KOGITO-2170
 - https://issues.redhat.com/browse/KOGITO-2213
 - https://issues.redhat.com/browse/KOGITO-2453

## Description
 - Integrated the state control API on the embedded-editor.
 - Added verification of the isDirty flag when the user tries to close the editor (online and desktop).
 - Created a package for the KeyboardShortcutsAPI.
 - Added ctrl+z (undo) and ctrl+shift+z (redo) keyboard shortcuts on the embedded-editor. (online, desktop, chrome extension)

### Demo
New isDirty indicator and new warning box for when an unsaved progress will be lost by closing the editor with the close button or with the logo button. (on Online and Desktop)
#### isDirty indicator - Online Editor
kogito-1337-online

####  isDirty indicator - Desktop
kogito-1337-desktop

#### Undo/Redo 
Use the shortcuts ctrl+z and ctrl+shift+z to respectively perform the Undo or Redo commands (on Online, Desktop and Chrome Extension)
ezgif com-video-to-gif

### CODE REVIEW

**Comment by ederign** [comment]
@porcelli ^

**Comment by manstis** [comment]
I'd appreciate it if this could wait for https://github.com/kiegroup/kogito-tooling/pull/142 to be merged.

The referenced PR moves most channels (not VSCode) to re-use a new generic _embedded_ editor rather than duplicate a lot of the envelope/bus code in different channels. I just have a couple more tests to write.

**Comment by manstis** [comment]
@ljmotta Hi.. your PR can proceed.... however (not unexpectedly) there are some conflicts now the re-usable embedded editor changes have been merged (`-desktop` and `-online` now use the same _core_ embedded editor so you will need to move some of your changes to `embedded-editor` but the code should look familiar, just in a different place). Also, whatever you change to `embedded-editor` will affect `-desktop`, `-online` and `-chrome-extension` so be warned!

**Comment by ederign** [comment]
@ljmotta  this needs a rebase!

**Comment by ljmotta** [comment]
@manstis Thank you for the heads up!

**Comment by ljmotta** [comment]
This PR isn't ready yet.

File: packages/microeditor-envelope/src/EnvelopeBusInnerMessageHandler.ts [hunk 1]
```diff
@@ -121,6 +122,9 @@ export class EnvelopeBusInnerMessageHandler {
   public respond_previewRequest(previewSvg: string) {
     return this.send({ type: EnvelopeBusMessageType.RETURN_PREVIEW, data: previewSvg });
   }
+  public request_stateControlCommandUpdateRequest(stateControl: StateControlCommand) {
```
Comment (tiagobento) [inline packages/microeditor-envelope/src/EnvelopeBusInnerMessageHandler.ts]: Rename this variable

File: packages/desktop/src/webview/editor/EditorToolbar.tsx [hunk 2]
```diff
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+import { useDirtyState } from "@kogito-tooling/embedded-editor";
```
Comment (tiagobento) [inline packages/desktop/src/webview/editor/EditorToolbar.tsx]: Unused import?

**Comment by manstis** [comment]
@Roberto FYI (related to our discussion)

File: packages/desktop/model.bpmn [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+              </bpsim:ProcessingTime>
+            </bpsim:TimeParameters>
+          </bpsim:ElementParameters>
+        </bpsim:Scenario>
+      </bpsim:BPSimData>
+    </bpmn2:extensionElements>
+    <bpmn2:source>_uIY5EJesEDiv7t8CormXnQ</bpmn2:source>
+    <bpmn2:target>_uIY5EJesEDiv7t8CormXnQ</bpmn2:target>
+  </bpmn2:relationship>
+</bpmn2:definitions>
```
Comment (tiagobento) [inline packages/desktop/model.bpmn]: Can we move this file to somewhere better? It's on the root of the package right now, but I guess it should be inside the tests directory?

File: packages/microeditor-envelope/src/EditorEnvelopeView.tsx [hunk 4]
```diff
@@ -17,19 +17,23 @@
 import * as React from "react";
 import * as ReactDOM from "react-dom";
 import * as AppFormer from "@kogito-tooling/core-api";
+import { ChannelType, StateControlCommand } from "@kogito-tooling/core-api";
 import { LoadingScreen } from "./LoadingScreen";
-import { KeyBindingsHelpOverlay } from "./KeyBindingsHelpOverlay";
-import { KeyboardShortcutsApi } from "./api/keyboardShortcuts";
+import { KeyBindingsHelpOverlay } from "@kogito-tooling/keyboard-shortcuts";
+import { KeyboardShortcutsApi } from "../../keyboard-shortcuts/src/api";
```
Comment (tiagobento) [inline packages/microeditor-envelope/src/EditorEnvelopeView.tsx]: This import is not correct

File: packages/microeditor-envelope/src/EditorEnvelopeView.tsx [hunk 5]
```diff
@@ -38,6 +42,9 @@ interface State {
 }
 
 export class EditorEnvelopeView extends React.Component<Props, State> {
+  private redoId: number;
+  private undoId: number;
```
Comment (tiagobento) [inline packages/microeditor-envelope/src/EditorEnvelopeView.tsx]: Can you rename those to `redoShortcutKeybindingId`? Same for undo.

File: packages/vscode-extension/src/KogitoEditor.ts [hunk 6]
```diff
@@ -144,6 +145,9 @@ export class KogitoEditor {
             const parsedPath = __path.parse(this.uri.fsPath);
             fs.writeFileSync(`${parsedPath.dir}/${parsedPath.name}-svg.svg`, preview);
           }
+        },
+        receive_stateControlCommandUpdate(stateControlCommand: StateControlCommand) {
+          /**/
```
Comment (tiagobento) [inline packages/vscode-extension/src/KogitoEditor.ts]: You can add a comment here saying that on VS Code, the command stack is handled by VS Code itself.

**Review by tiagobento — CHANGES_REQUESTED** [review]
Minor change requests inline. Thanks @ljmotta!

File: packages/desktop/src/webview/editor/EditorToolbar.tsx (hunk 2 above)
Comment (ljmotta) [reply packages/desktop/src/webview/editor/EditorToolbar.tsx]: Yes. Already removed.

File: packages/desktop/model.bpmn (hunk 3 above)
Comment (ljmotta) [reply packages/desktop/model.bpmn]: My bad! This file wasn't suppose to be committed, it was generated on one of my manual tests.

File: packages/microeditor-envelope/src/EditorEnvelopeView.tsx (hunk 4 above)
Comment (ljmotta) [reply packages/microeditor-envelope/src/EditorEnvelopeView.tsx]: Done

File: packages/microeditor-envelope/src/EditorEnvelopeView.tsx (hunk 5 above)
Comment (ljmotta) [reply packages/microeditor-envelope/src/EditorEnvelopeView.tsx]: Done

File: packages/vscode-extension/src/KogitoEditor.ts (hunk 6 above)
Comment (ljmotta) [reply packages/vscode-extension/src/KogitoEditor.ts]: Done

**Review by tiagobento — APPROVED** [review]

**Comment by domhanak** [comment]
Hi @ljmotta I did exploratory testing on Desktop Channel, see below

Desktop tests
isDirty -In general Works well - Issue I found:
in DMN when I added Documentation Link to the node, the file was not marked as edited and I was able to close the editor and lost the data I inputted. @jomarko Maybe it is because this property is not stored in the DMN file? Imho this is not severe, and we can report it as a bug.

stateControl - In general Works Well - Issue I found:
in DMN when I added Documentation Link to the node, I wasn't able to do undo and redo - same as above

Common issue - when you edit properties and do UNDO, file stays in EDITED state, however when you edit canvas (move nodes, change names etc.) a do UNDO, the file is moved back and - EDITED is not shown in title. I think this should be consistent, so when I edit property and do 1 UNDO, I no longer have edited state of the file.
Workaround (not 100% of the time) - do the UNDO twice when changing something in property panel in BPMN, DMN editor needs only one.

Common issue in both Editors:
1. Double click the node on canvas and change it's name and confirm by ENTER (do not click the check-mark)
2. Do UNDO
3. Name of the node is not reverted back
Workaround: Use properties panel to update names OR click the check mark to update the name on canvas and UNDO will work.

BPMN issue:
1. Change process type  Private -> public
2. File is EDITED.
3. Do UNDO once or twice
4. File is not edited but Process type remains changed - Public


DMN Editor supports DataTypes - does this PR affect those aswell? I added a new DataType but there was no EDITED state displayed.

**Comment by domhanak** [comment]
Issue with annotations in DMN
1. Add annotation - name ends with `....notation-1`
2. Do UNDO
3. DO REDO
4. name of the annotation is changed - ends with `....notation-2`
If you continue doing undo/redo the name will always be an increment of previous one

**Comment by ljmotta** [comment]
Thanks for the review @domhanak !

1) Documentation Link
I confirmed here, and as I could see, the editor doesn't send a message to the channel when a Documentation Link is inserted/removed, so It's not possible to keep track of then with the StateControlAPI. This bug is Editor related (checked with the published version of the VS Code extension). Created a JIRA for this bug https://issues.redhat.com/browse/KOGITO-2555

2) File keep the Edited indicator after a `undo` operation
For now the KeybordShortcuts are being added only on the canvas. So if you edit a property and try to undo it without selecting the canvas, the **text** undo/redo will take action (not the StateControlAPI), and because of that, it will be a new edition. After selecting the canvas it'll be possible to undo/redo the editions on the properties. I will wait for https://github.com/kiegroup/kogito-tooling/pull/185 to be merge and then see what can be done.

3) Enter
This is another bug related to the Editor itself (checked with the published version of the VS Code extension). Created a JIRA for this bug https://issues.redhat.com/browse/KOGITO-2553

4) Private -> Public
This is another bug related to the Editor itself (checked with the published version of the VS Code extension). Seems that the editor doesn't handle the undo/redo command for the process type. Created a JIRA for this bug https://issues.redhat.com/browse/KOGITO-2554

5) Annotations
This is another bug related to the Editor itself (checked with the published version of the VS Code extension). Probably the editor keeps count of how many notations was created, and aren't updated with the undo/redo command. Created a JIRA for this bug https://issues.redhat.com/browse/KOGITO-2557

**Comment by domhanak** [comment]
@ljmotta Thanks for confirmation, I suspected it is editor related. :) I ll only need to check the chrome-extension now.

**Comment by ljmotta** [comment]
@domhanak  After resolving the conflicts with the https://github.com/kiegroup/kogito-tooling/pull/185 I tested again the 2) case. Appears that this is another bug related with the Editor. The same behavior occurs on the VS Code extension. I created a JIRA (https://issues.redhat.com/browse/KOGITO-2559) to keep track of this problem.
If any problem appears on the chrome-extension let me know :slightly_smiling_face:

**Comment by domhanak** [comment]
@ljmotta Thanks for reporting. I tried chrome-extension, I didn't see anything broken, but when I try to edit the file on GitHub - (example):
https://github.com/kiegroup/kogito-examples/edit/stable/kogito-travel-agency/basic/src/main/resources/org/acme/travels/flightBooking.bpmn2 
I can't use UNDO/REDO is that expected?

Use case would be:
1. Open file on GH, it says this is read-only mode and if I do change UNDO/REDO does not work (this looks ok to me)
2. Click Edit (pen icon on the top-left)
3. Do a change on canvas, try doing UNDO or REDO
4. Nothing happens

From the JIRA and description in PR I expected this will work. Let me know if it is supported like this or user should always go to online editor, make changes there and come back to GH. Which imho, adds unnecessary  clicks for users.

Edit: I tried fullscreen too but it did not help

**Comment by ljmotta** [comment]
@domhanak Seems like the chrome extension is using the released editor. I suppose that within the Undo/Redo functionalities none of the KeyboardShortcuts features was enabled as well?

To test this PR it's necessary to use `yarn build:fast`, it'll make the chrome extension use the local changes and not use the editor on GitHub.

**Comment by domhanak** [comment]
@ljmotta well ok I ll try locally again, is there a way to fix the build so  that the artifact for chrome extension contains changes from the PR?

About the shortcuts, ENTER< DELETE worked as usual, only UNDO/REDO are not doing anything

**Review by domhanak — APPROVED** [review]
LGTM.
Thanks @ljmotta great job.

Worked well on all related channels, issues found are not related to this PR and each of them is reported under editor component in JIRA.

**Review by ederign — APPROVED** [review]

**Comment by ljmotta** [comment]
@domhanak About create a chrome extension artifact with changes regarding the editors:
The editors weren't shipped with the chrome extension, so this wouldn't remove the necessity to run the local commands. With that in mind, I can't see the benefits to add this new artifact.

**Pull request merged**